### PR TITLE
Override the print macros with a version suitable for verification

### LINKED ISSUE #194: Incorrect verification success on print statement
RMC run with bound checking says verification is successful on this function, which actually seg faults.

Command: `rmc --keep-temps --gen-c src/test/cbmc/DynTrait/fixme_offset.rs --pointer-check --bounds-check`

```rust
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0 OR MIT

struct Foo {}
trait Trait {
    fn func(&self) -> i32;
}

impl Trait for Foo {
    fn func(&self) -> i32 {
        1 as i32
    }
}

fn foo() -> Box<dyn Trait> {
    Box::new(Foo {}) 
}

fn main() {
    let f = foo();
    unsafe {
        let f_ptr = Box::into_raw(f) as *const usize;
        println!("{}", *(f_ptr.offset(100000)));
    }
}
```

RMC result:
```
/Users/avanhatt/research/rmc/src/test/cbmc/DynTrait/fixme_offset.rs function main
[main.assertion.3] line 20 resume instruction: SUCCESS
[main.assertion.1] line 22 Non-null virtual function call for "::Trait::func": SUCCESS
[main.pointer_dereference.1] line 22 dereference failure: pointer NULL in var_3.vtable->::Trait::func: SUCCESS
[main.pointer_dereference.2] line 22 dereference failure: pointer invalid in var_3.vtable->::Trait::func: SUCCESS
[main.pointer_dereference.3] line 22 dereference failure: deallocated dynamic object in var_3.vtable->::Trait::func: SUCCESS
[main.pointer_dereference.4] line 22 dereference failure: dead object in var_3.vtable->::Trait::func: SUCCESS
[main.pointer_dereference.5] line 22 dereference failure: pointer outside object bounds in var_3.vtable->::Trait::func: SUCCESS
[main.pointer_dereference.6] line 22 dereference failure: invalid integer address in var_3.vtable->::Trait::func: SUCCESS
[main.assertion.2] line 23 assertion failed: x == 1: SUCCESS

<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

** 0 of 78 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

Rust result:
```
timeout: the monitored command dumped core
/playground/tools/entrypoint.sh: line 11:     9 Segmentation fault      timeout --signal=KILL ${timeout} "$@"
```

### PR DESCRIPTION
**Pull request by zhassan-aws**
### Description of changes: 

Currently any programs containing (reachable) print macros (e.g. `println`) result in verification failure (see #824). This PR overrides std's print macros by providing versions that are suitable for verification.

### Resolved issues:

Resolves #194 and #824 

### Call-outs:

The overriden versions of the macros do not properly parse the arguments, so Kani may accept programs that are rejected by rustc, e.g.
```rust
fn main() {
    println!("Two arguments expected, but only one provided: {} {}", 4+5);
}
```

### Testing:

* How is this change tested? Added two new tests.

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: library/std/src/lib.rs [hunk 1]
```diff
@@ -45,3 +45,38 @@ macro_rules! assert {
         kani::assert($cond, concat!(stringify!($($arg)+)));
     };
 }
+
+#[macro_export]
+macro_rules! evaluate_print_args {
+    () => { /* do nothing */ };
+    ($x:expr $(, $arg:expr)* $(,)?) => {
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply library/std/src/lib.rs]: Fixed and added a comment to clarify.

File: library/std/src/lib.rs [hunk 2]
```diff
@@ -45,3 +45,38 @@ macro_rules! assert {
...
+        {
+            $(
+                $arg;
+            )*
+        }
+    };
+}
+
+// Override the print macros to skip all the formatting functionality (which
+/// is not relevant for verification)
```
[the remark this answers is not in the document]
Comment (zhassan-aws) [reply library/std/src/lib.rs]: Fixed.

File: tests/kani/Print/side_effects.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This test checks that Kani's modeling of println and eprintln takes into
```
Comment (adpaco-aws) [inline tests/kani/Print/side_effects.rs:4]:
```suggestion
// This test checks that Kani's modeling of print macros takes into
```

File: tests/kani/Print/side_effects.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    y
+}
+
+fn main() {
+    let mut x = 5;
+    println!("calling function with side-effect from println!: {}", subtract_two(&mut x));
+    assert!(x == 10);
+
+    eprintln!("calling function with side-effect from eprintln!: {}", subtract_two(&mut x));
+    assert!(x == 15);
```
Comment (adpaco-aws) [inline tests/kani/Print/side_effects.rs:20]: Why not test the "no newline" variants here as well?

File: tests/kani/Print/main.rs [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+// This test checks that println! and eprintln! do not result in verification
```
Comment (adpaco-aws) [inline tests/kani/Print/main.rs:4]:
```suggestion
// This test checks that `print!`, `println!`, `eprint!` and `eprintln!` do not result in verification
```
or just use "print macros" to refer to these

**Comment by zhassan-aws** [comment]
> Really excited about this, thank you @zhassan-aws !!
> 
> My only concern about this is the lack of user-facing documentation. What do you think about keeping a list or table of the macros/functions we override that describes the behavior or limitations for each one (e.g., the message formatting call-out in these print macros).

Yes, this makes sense. Should I put this under [Limitations section](https://model-checking.github.io/kani/limitations.html ) of the docs?

**Comment by adpaco-aws** [comment]
Yes, I would either add it to that section or create a subsection under limitations for overrides specifically. Both options are fine with me for now, we can define a better structure once we review the documentation.

File: tests/kani/Print/main.rs (hunk 5 above)
Comment (zhassan-aws) [reply tests/kani/Print/main.rs]: Fixed.

**Comment by zhassan-aws** [comment]
> Yes, I would either add it to that section or create a subsection under limitations for overrides specifically. Both options are fine with me for now, we can define a better structure once we review the documentation.

Done.

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**